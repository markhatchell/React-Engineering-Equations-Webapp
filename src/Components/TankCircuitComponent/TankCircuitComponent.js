import React from 'react';
import BaseEquation from '../BaseEquation';
import { TankCircuit } from 'js-electrical-engineering-equations';
import '../../Styles/Equations.css';

// equations https://www.allaboutcircuits.com/textbook/alternating-current/chpt-6/parallel-tank-circuit-resonance/


class TankCircuitComponent extends BaseEquation {

  constructor(props) {
    super(props);
    this.state = {
      frequencyInHertz: '',
      inductanceInNanoHenries: '',
      capacitanceInPicoFarads: '',
      watts: '',
      lastCalculated: []
    }
  }

  render() {
    return (
      <div className="equationBox">
        <h3>Tank Circuit</h3>
        <div className="equationBoxInner">
          <p>Supply at least two of the values to calculate the third.</p>
          <div className="tableContainer">
            <form onSubmit={(e) => this.handleSubmit(e)}>
              <table className="table">
                <tbody>
                <tr>
                  <td align="left">
                    <label htmlFor="tankcircuit-F"> &#x192; (Frequency in Hertz)</label>
                  </td>
                  <td>
                    <input id="tankcircuit-F" value={this.state.frequencyInHertz} disabled={this.wasLastCalculated('frequencyInHertz')}
                           autoComplete="off"
                           tabIndex="1"
                           onChange={(e) => this.collectFloatValueFor('frequencyInHertz', e)}/>
                  </td>
                </tr>
                <tr>
                  <td align="left">
                    <label htmlFor="tankcircuit-L">L (Inductance in nH)</label>
                  </td>
                  <td>
                    <input id="tankcircuit-L" value={this.state.inductanceInNanoHenries} disabled={this.wasLastCalculated('inductanceInNanoHenries')}
                           autoComplete="off"
                           tabIndex="2"
                           onChange={(e) => this.collectFloatValueFor('inductanceInNanoHenries', e)}/>
                  </td>
                </tr>
                <tr>
                  <td align="left">
                    <label htmlFor="tankcircuit-C">C (Capacitance in pF)</label>
                  </td>
                  <td>
                    <input id="tankcircuit-C" value={this.state.capacitanceInPicoFarads} disabled={this.wasLastCalculated('capacitanceInPicoFarads')}
                           autoComplete="off"
                           tabIndex="3"
                           onChange={(e) => this.collectFloatValueFor('capacitanceInPicoFarads', e)}/>
                  </td>
                </tr>
                <tr>
                  <td align="left">
                    <button type="button" tabIndex="5" onClick={() => this.handleClear()}>Clear</button>
                  </td>
                  <td align="right">
                    <button type="submit" tabIndex="4">Calculate</button>
                  </td>
                </tr>
                </tbody>
              </table>
            </form>
          </div>
        </div>
      </div>
    );
  }

  calculateOtherValues() {
    const { frequencyInHertz, inductanceInNanoHenries ,capacitanceInPicoFarads } = this.state;
    const inductance  = parseFloat(this.state.inductanceInNanoHenries) / 1000000000;
    const capacitance = parseFloat(this.state.capacitanceInPicoFarads) / 1000000000000;


    if (this.wasLastCalculated('frequencyInHertz') || (inductanceInNanoHenries && capacitanceInPicoFarads && frequencyInHertz === '')) {
      const { calculatedFrequencyInHertz } = TankCircuit.calcFrequency(inductance, capacitance);
      this.setState(
        Object.assign(
          {},
          this.state,
          {
            frequencyInHertz: numberWithCommas(calculatedFrequencyInHertz),
            lastCalculated: ['frequencyInHertz']
          }
        )
      );
    } else if (this.wasLastCalculated('capacitanceInPicoFarads') || (frequencyInHertz && inductanceInNanoHenries && capacitanceInPicoFarads === '')) {
      const frequencyAsFloat = parseFloat(frequencyInHertz.replace(/,/g,''));
      const { calculatedCapacitanceInFarads } = TankCircuit.calcCapacitance(frequencyAsFloat, inductance);

      this.setState(
        Object.assign(
          {},
          this.state,
          {
            capacitanceInPicoFarads: calculatedCapacitanceInFarads * 1000000000000,
            lastCalculated: ['capacitanceInPicoFarads']
          }
        )
      );
    } else if (this.wasLastCalculated('inductanceInNanoHenries') || (frequencyInHertz && capacitanceInPicoFarads && inductanceInNanoHenries === '')) {
      const frequencyAsFloat = parseFloat(frequencyInHertz.replace(/,/g,''));
      const { calculatedInductanceInHenries } = TankCircuit.calcInductance(frequencyAsFloat, capacitance);

      this.setState(
        Object.assign(
          {},
          this.state,
          {
            inductanceInNanoHenries: calculatedInductanceInHenries * 1000000000,
            lastCalculated: ['inductanceInNanoHenries']
          }
        )
      );
    }

  }

  countValues() {
    const values = [this.state.frequencyInHertz, this.state.inductanceInNanoHenries, this.state.capacitanceInPicoFarads];
    let valuesCount = 0;
    for (let i = 0; i < values.length; i++) {
      if (values[i] !== '') {
        valuesCount++;
      }
    }
    return valuesCount;
  }

  handleSubmit(e) {
    const {lastCalculated} = this.state;
    if (this.countValues() === 2 || lastCalculated !== []) {
      this.calculateOtherValues();
    } else {
      this.clear();
    }
    e.preventDefault();
    return false;
  }

  clear() {
    this.setState(
      Object.assign(
        {},
        this.state,
        {
          frequencyInHertz: '',
          inductanceInNanoHenries: '',
          capacitanceInPicoFarads: '',
          lastCalculated: ''
        }
      )
    );
  }


  handleClear() {
    this.clear();
  }

  wasLastCalculated(field) {
    const lastCalculated = this.state.lastCalculated;
    return lastCalculated.indexOf(field) >= 0;

  }

}

const numberWithCommas = (x) => {
  let sides = x.toString().split('.');
  if (!sides.hasOwnProperty(1)) {
    sides[1] = 0;
  }
  return sides[0].replace(/\B(?=(\d{3})+(?!\d))/g, ",") + '.' +  sides[1];
};


export default TankCircuitComponent;