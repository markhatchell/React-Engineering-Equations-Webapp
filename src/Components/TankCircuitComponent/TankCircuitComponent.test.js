import React from 'react';
import renderer from 'react-test-renderer';
import TankCircuitComponent from "./TankCircuitComponent";
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

const dom = renderer.create(
  <TankCircuitComponent />,
);

describe('TankCircuitComponent', () => {

  test('snapshot matches', () => {
    let tree = dom.toJSON();
    expect(tree).toMatchSnapshot();
  });

  const component = shallow(<TankCircuitComponent />);
  const instance = component.instance();


});
