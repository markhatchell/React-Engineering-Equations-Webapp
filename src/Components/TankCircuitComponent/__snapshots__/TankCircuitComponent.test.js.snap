// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TankCircuitComponent snapshot matches 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation2-6 MuiPaper-rounded-3 MuiCard-root-1 equationBox"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiCardContent-root-29 equationBoxInner"
    >
      <p>
        Supply at least two of the values to calculate the third.
      </p>
      <div
        aria-describedby={undefined}
        className="MuiFormControl-root-30 MuiFormControl-marginNormal-31 MuiFormControl-fullWidth-33"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        <label
          className="MuiFormLabel-root-39 MuiInputLabel-root-34 MuiInputLabel-formControl-35 MuiInputLabel-animated-38 MuiInputLabel-shrink-37"
          data-shrink={true}
          htmlFor="full-width"
        >
          Æ’ (Frequency in Hertz)
        </label>
        <div
          className="MuiInput-root-44 MuiInput-fullWidth-51 MuiInput-formControl-45 MuiInput-underline-48"
        >
          <input
            aria-invalid={false}
            aria-required={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-52"
            defaultValue={undefined}
            disabled={false}
            id="full-width"
            name="frequencyInHertz"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        aria-describedby={undefined}
        className="MuiFormControl-root-30 MuiFormControl-marginNormal-31 MuiFormControl-fullWidth-33"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        <label
          className="MuiFormLabel-root-39 MuiInputLabel-root-34 MuiInputLabel-formControl-35 MuiInputLabel-animated-38 MuiInputLabel-shrink-37"
          data-shrink={true}
          htmlFor="full-width"
        >
          L (Inductance in nH)
        </label>
        <div
          className="MuiInput-root-44 MuiInput-fullWidth-51 MuiInput-formControl-45 MuiInput-underline-48"
        >
          <input
            aria-invalid={false}
            aria-required={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-52"
            defaultValue={undefined}
            disabled={false}
            id="full-width"
            name="inductanceInNanoHenrys"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        aria-describedby={undefined}
        className="MuiFormControl-root-30 MuiFormControl-marginNormal-31 MuiFormControl-fullWidth-33"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        <label
          className="MuiFormLabel-root-39 MuiInputLabel-root-34 MuiInputLabel-formControl-35 MuiInputLabel-animated-38 MuiInputLabel-shrink-37"
          data-shrink={true}
          htmlFor="full-width"
        >
          C (Capacitance in pF)
        </label>
        <div
          className="MuiInput-root-44 MuiInput-fullWidth-51 MuiInput-formControl-45 MuiInput-underline-48"
        >
          <input
            aria-invalid={false}
            aria-required={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-52"
            defaultValue={undefined}
            disabled={false}
            id="full-width"
            name="capacitanceInPicoFarads"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiCardActions-root-57"
    >
      <button
        className="MuiButtonBase-root-74 MuiButton-root-59 MuiCardActions-action-58"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="5"
        type="button"
      >
        <span
          className="MuiButton-label-60"
        >
          Clear
        </span>
        <span
          className="MuiTouchRipple-root-77"
        />
      </button>
      <button
        className="MuiButtonBase-root-74 MuiButton-root-59 MuiButton-raised-64 MuiButton-raisedPrimary-65 MuiCardActions-action-58"
        disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "marginLeft": "auto",
          }
        }
        tabIndex="4"
        type="submit"
      >
        <span
          className="MuiButton-label-60"
        >
          Calculate
        </span>
        <span
          className="MuiTouchRipple-root-77"
        />
      </button>
    </div>
  </form>
</div>
`;
